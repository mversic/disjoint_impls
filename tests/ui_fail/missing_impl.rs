use disjoint_impls::disjoint_impls;

pub trait Dispatch1 {
    type Group;
}
pub trait Dispatch2 {
    type Group;
}

pub enum GroupA {}
pub enum GroupB {}

impl Dispatch1 for String {
    type Group = GroupA;
}
impl Dispatch2 for String {
    type Group = GroupA;
}
impl<T> Dispatch1 for Vec<T> {
    type Group = GroupA;
}
impl<T> Dispatch2 for Vec<T> {
    type Group = GroupB;
}

impl Dispatch1 for i32 {
    type Group = GroupB;
}
impl Dispatch1 for u32 {
    type Group = GroupB;
}

disjoint_impls! {
    pub trait Kita {
        const NAME: &'static str;
    }

    impl<T: Dispatch1<Group = GroupA> + Dispatch2<Group = GroupA>> Kita for T {
        const NAME: &'static str = "Blanket A";
    }
    impl<T: Dispatch1<Group = GroupB>> Kita for T {
        const NAME: &'static str = "Blanket B";
    }
}

fn main() {
    assert_eq!("Blanket A", String::NAME);
    assert_eq!("Blanket B", u32::NAME);
    assert_eq!("Blanket B", i32::NAME);

    let _ = <Vec::<u32> as Kita>::NAME;
}

