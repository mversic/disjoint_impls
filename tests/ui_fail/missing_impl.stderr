error[E0271]: type mismatch resolving `<Vec<u32> as Dispatch2>::Group == GroupA`
  --> tests/ui_fail/missing_impl.rs:51:14
   |
51 |     let _ = <Vec::<u32> as Kita>::NAME;
   |              ^^^^^^^^^^ type mismatch resolving `<Vec<u32> as Dispatch2>::Group == GroupA`
   |
note: expected this to be `GroupA`
  --> tests/ui_fail/missing_impl.rs:23:18
   |
23 |     type Group = GroupB;
   |                  ^^^^^^
note: required for `Vec<u32>` to implement `Kita0<GroupA>`
  --> tests/ui_fail/missing_impl.rs:33:1
   |
33 | / disjoint_impls! {
34 | |     pub trait Kita {
35 | |         const NAME: &'static str;
...  |
44 | | }
   | |_^
note: required for `Vec<u32>` to implement `Kita`
  --> tests/ui_fail/missing_impl.rs:33:1
   |
33 | / disjoint_impls! {
34 | |     pub trait Kita {
35 | |         const NAME: &'static str;
...  |
38 | |     impl<T: Dispatch1<Group = GroupA> + Dispatch2<Group = GroupA>> Kita for T {
   | |                                                                    ^^^^
...  |
44 | | }
   | |_^
   = note: this error originates in the macro `disjoint_impls` (in Nightly builds, run with -Z macro-backtrace for more info)
